### DISCUSSION TITLE
How to configure the sme vector length in arm extension?

### DISCUSSION THREAD

**Discussion by hanm2019** [gem5/gem5#1791] [Q&A] [answered]
I am currently exploring the performance of the SME extension under different SME vector length configurations using gem5. However, I have not been able to find clear instructions on how to enable the SME extension or configure the SME vector length. I would greatly appreciate your guidance on this matter.  

To provide some context on my setup:  
1. I created a folder named `matrix` under the `tests` directory.  
2. Inside, I implemented a matrix multiplication program, `matrix.c`.  
3. I wrote a `Makefile` to compile the program using Clang.  
4. run `make matrix` and `make run_matrix`

[source.zip](https://github.com/user-attachments/files/17824129/source.zip)

**Comment by giactra** [comment, marked as answer]
Hi! Thanks for your question.

As of now FEAT_SME is enabled by default.

The way you change the VL depends on the running mode:

1) FS = FullSystem
You can change the VL with the **sme_vl** parameter in the ArmSystem: https://github.com/gem5/gem5/blob/stable/src/arch/arm/ArmSystem.py#L326

2) SE = SyscallEmulation
You can change the VL with the **sme_vl_se** parameter in the ArmISA: https://github.com/gem5/gem5/blob/stable/src/arch/arm/ArmISA.py#L190

The reason for this differentiation is that SE mode does not have an ArmSystem.

Bear in mind sme_vl indicates the vector length in quadwords (128-bit) so for example with a 256bits VL you should set it to 2.

**Comment by hanm2019** [comment]
Thank you for your response!
I have another question: when using SE mode, the code
```
    printf("Has SME? %s\n", __arm_has_sme() ? "Yes" : "No");
```
report `Has SME? No`

The `__arm_has_sme() ` seems a intrinsic implemented by clang

```
__ai bool __arm_has_sme(void) __arm_streaming_compatible {
  uint64_t x0, x1;
  __builtin_arm_get_sme_state(&x0, &x1);
  return x0 & (1ULL << 63);
}

```

**Reply by giactra** [reply]
Thanks for spotting this. Let me know if https://github.com/gem5/gem5/pull/1792 fixes the problem
